components/fxa-client/src/pairing_channel/messages.rs

```diff
@@ -0,0 +1,672 @@
+
+impl From<ClientHello> for HandshakeMessage {
+    fn from(client_hello: ClientHello) -> Self {
+        Self {
+            message_type: client_hello.get_type_tag().into(),
+            message: Box::new(client_hello),
+        }
+    }
+}
+
+impl ClientHello {
+    #[allow(dead_code)]
+    pub fn new(random: Vec<u8>, session_id: Vec<u8>, extensions: Vec<Extension>) -> Self {
+        Self {
+            random,
+            session_id,
+            extensions,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn get_session_id(&self) -> Vec<u8> {
+        self.session_id.clone()
+    }
+
+    #[allow(dead_code)]
+    pub fn get_extension(&self, ext_num: u16) -> Option<&Extension> {
+        self.extensions
+            .iter()
+            .find(|ext| ext.get_type_tag() == ext_num)
+    }
+
+    fn read(buf: &mut Bytes) -> Result<Self> {
+        // The legacy_version field may indicate an earlier version of TLS
+        // for backwards compatibility, but must not predate TLS 1.0!
+        let version = utils::read_u16(buf)?;
+        if version < VERSION_TLS_1_0 {
+            anyhow::bail!("Invalid protocol version")
+        }
+        let len = 32;
+        let random = utils::read_bytes(buf, len as usize)?;
+        let session_id = utils::read_bytes_with_u8_len(buf)?;
+        let mut found = false;
+        let inner_data = utils::read_bytes_with_u16_len(buf)?;
+        let mut inner_data_buf = inner_data.as_slice().to_bytes();
+        // We only support a single ciphersuite, but the peer may offer several.
+        // Scan the list to confirm that the one we want is present.
+        while inner_data_buf.has_remaining() {
+            let cipher_suite = utils::read_u16(&mut inner_data_buf)?;
+            if cipher_suite == TLS_AES_128_GCM_SHA256 {
+                found = true;
+            }
+        }
+        if !found {
+            anyhow::bail!("Handshake failure")
+        }
+        // legacy_compression_methods must be a single zero byte for TLS1.3 ClientHellos.
+        // It can be non-zero in previous versions of TLS, but we're not going to
+        // make a successful handshake with such versions, so better to just bail out now.
+        let legacy_compression_modes = utils::read_bytes_with_u8_len(buf)?;
+        if legacy_compression_modes.len() != 1 || legacy_compression_modes[0] != 0x00 {
+            anyhow::bail!("Illegal prameter for compression modes")
+        }
+        let (last_seen, extensions) =
+            HandshakeMessage::read_extensions(HandshakeType::ClientHello, buf)?;
+
+        if !extensions
+            .iter()
+            .find(|ext| ext.get_type_tag() == SUPPORTED_VERSIONS_ID)
+            .ok_or_else(|| anyhow::Error::msg("Missing supported versions"))?
+            .contains_version(VERSION_TLS_1_3)
+        {
+            anyhow::bail!("Does not support TLS1.3")
+        }
+
+        if extensions
+            .iter()
+            .any(|ext| ext.get_type_tag() == PRE_SHARED_KEY_ID)
```

Comment:
Naively, I would have expected to see this matching on the `ExtensionType::PreSharedKey` enum rather than a numerical constant, so I'm curious if you tried that and whether it turned out to be awkward in practice. (You might have to derive `Copy` and `Eq`, or make the enum `repr(u16)` or something in order to get the nice `==` matching behaviour though).

Reply:
Aaah ya that's leftover from when I had the extension be a Map and the key was a u16, I'll change it up (I believe I do similar stuff with Messages as well, so I'll get that done too)